src/main/java/dev/openfeature/sdk/ProviderRepository.java

```diff
@@ -0,0 +1,156 @@
+        initializingDefaultProvider = provider;
+        initializeProvider(provider, this::updateDefaultProviderAfterInitialization);
+    }
+
+    private void initializeProvider(String clientName, FeatureProvider provider) {
+        initializingNamedProviders.put(clientName, provider);
+        initializeProvider(provider, newProvider -> updateProviderAfterInit(clientName, newProvider));
+    }
+
+    private void initializeProvider(FeatureProvider provider, Consumer<FeatureProvider> afterInitialization) {
+        taskExecutor.submit(() -> {
+            try {
+                if (isProviderNotBoundMultipleTimes(provider)) {
+                    provider.initialize();
+                }
+                afterInitialization.accept(provider);
+            } catch (Exception e) {
+                log.error("Exception when initializing feature provider {}", provider.getClass().getName(), e);
+            }
+        });
+    }
+
+    private void updateProviderAfterInit(String clientName, FeatureProvider newProvider) {
+        Optional
+                .ofNullable(initializingNamedProviders.get(clientName))
+                .filter(initializingProvider -> initializingProvider.equals(newProvider))
+                .ifPresent(provider -> updateNamedProviderAfterInitialization(clientName, provider));
+    }
+
+    private void updateDefaultProviderAfterInitialization(FeatureProvider initializedProvider) {
+        Optional
+                .ofNullable(this.initializingDefaultProvider)
+                .filter(initializingProvider -> initializingProvider.equals(initializedProvider))
+                .ifPresent(this::replaceDefaultProvider);
+    }
+
+    private void replaceDefaultProvider(FeatureProvider provider) {
+        FeatureProvider oldProvider = this.defaultProvider.getAndSet(provider);
+        if (isOldProviderNotBoundByName(oldProvider)) {
+            shutdownProvider(oldProvider);
+        }
+    }
+
+    private boolean isOldProviderNotBoundByName(FeatureProvider oldProvider) {
+        return !this.providers.containsValue(oldProvider);
+    }
+
+    private void updateNamedProviderAfterInitialization(String clientName, FeatureProvider initializedProvider) {
+        Optional
+                .ofNullable(this.initializingNamedProviders.get(clientName))
+                .filter(initializingProvider -> initializingProvider.equals(initializedProvider))
+                .ifPresent(provider -> replaceNamedProviderAndShutdownOldOne(clientName, provider));
+    }
+
+    private void replaceNamedProviderAndShutdownOldOne(String clientName, FeatureProvider provider) {
+        FeatureProvider oldProvider = this.providers.put(clientName, provider);
+        this.initializingNamedProviders.remove(clientName, provider);
+        if (isProviderNotBoundMultipleTimes(oldProvider)) {
+            shutdownProvider(oldProvider);
+        }
+    }
+
+    private boolean isProviderNotBoundMultipleTimes(FeatureProvider oldProvider) {
```

Comment:
Minor remark - Here the logic is to check whether the provider is already registred either as default or as a named provider. Given that, IMO renaming it to something like `isProviderRegistered` suits here.

Reply:
i liked the not in the method name more than the `!` in the if conditions. but it's fine :)